lib/browse_everything/driver/dropbox.rb

```diff
@@ -1,63 +1,99 @@
+      class FileEntryFactory
+        def self.build(metadata:, key:)
+          factory_klass = klass_for metadata
+          factory_klass.build(metadata: metadata, key: key)
+        end
+
+        class << self
+          private def klass_for(metadata)
+            case metadata
+            when DropboxApi::Metadata::File
+              FileFactory
+            else
+              ResourceFactory
+            end
+          end
+        end
+      end
+
+      class ResourceFactory
+        def self.build(metadata:, key:)
+          path = metadata.path_display
+          BrowseEverything::FileEntry.new(
+            path,
+            [key, path].join(':'),
+            File.basename(path),
+            nil,
+            nil,
+            true
+          )
+        end
+      end
+
+      class FileFactory
+        def self.build(metadata:, key:)
+          path = metadata.path_display
+          BrowseEverything::FileEntry.new(
+            path,
+            [key, path].join(':'),
+            File.basename(path),
+            metadata.size,
+            metadata.client_modified,
+            false
+          )
+        end
+      end
 
       def icon
         'dropbox'
       end
 
       def validate_config
-        return if CONFIG_KEYS.all? { |key| config[key].present? }
-        raise BrowseEverything::InitializationError, "Dropbox driver requires #{CONFIG_KEYS.inspect}"
+        raise BrowseEverything::InitializationError, 'Dropbox driver requires a :client_id argument' unless config[:client_id]
+        raise BrowseEverything::InitializationError, 'Dropbox driver requires a :client_secret argument' unless config[:client_secret]
       end
 
-      # @return [Array<BrowseEverything::FileEntry>]
       def contents(path = '')
-        path.sub!(%r{ /^[\/.]+/}, '')
-        result = add_directory_entry(path)
-        result += client.metadata(path)['contents'].collect { |info| make_file_entry(info) }
-        result
+        result = client.list_folder(path)
+        result.entries.map { |entry| FileEntryFactory.build(metadata: entry, key: key) }
       end
 
-      def add_directory_entry(path)
-        return [] if path.empty?
-        [BrowseEverything::FileEntry.new(
-          Pathname(path).join('..'),
-          '', '..', 0, Time.zone.now, true
-        )]
+      def details(path)
+        metadata = client.get_metadata(path)
+        FileEntryFactory.build(metadata: metadata, key: key)
       end
 
-      def make_file_entry(info)
-        path = info['path']
-        BrowseEverything::FileEntry.new(
-          path,
-          [key, path].join(':'),
-          File.basename(path),
-          info['bytes'],
-          Time.zone.parse(info['modified']),
-          info['is_dir']
-        )
+      def download(path)
+        temp_file = Tempfile.open(File.basename(path), encoding: 'ascii-8bit')
+        client.download(path) do |chunk|
+          temp_file.write chunk
+        end
+        temp_file.close
+        temp_file
```

Comment:
I think this can be simplified to:
``` ruby
def download(path)
  Tempfile.open(File.basename(path), encoding: 'ascii-8bit') do |temp_file|
    client.download(path) do |chunk|
      temp_file.write chunk
    end
  end
end
```